tests/zfs-tests/tests/functional/cli_root/zfs_send/zfs_send_007_pos.ksh

```diff
@@ -63,15 +58,17 @@ test_pool ()
 	log_must zfs create -o recordsize=512 $POOL/fs
 	mntpnt=$(get_prop mountpoint "$POOL/fs")
 	log_must dd if=/dev/urandom of=${mntpnt}/file bs=512 count=1 2>/dev/null
-	first_object=$(ls -i $mntpnt | awk '{print $1}')
+	object=$(ls -i $mntpnt | awk '{print $1}')
 	log_must zfs snapshot $POOL/fs@a
 	while true; do
-		log_must find $mntpnt/* -delete
+		log_must find $mntpnt/ -type f -delete
 		sync
 		log_must mkfiles "$mntpnt/" 4000
-		FILE=$(ls -i $mntpnt | awk \
-			'{if ($1 == '$first_object') {print $2}}')
-		if [[ -n "$FILE" ]]; then
+		sync
+		# check if we started reusing objects
+		object=$(ls -i $mntpnt | sort -n | awk -v object=$object \
+		    '{if ($1 <= object) {exit 1}} END {print $1}')
```

Comment:
Yes, from my (limited) understanding the issue is caused by `large_dnodes` + `multi-threaded dmu_object_alloc()`: ever since large_dnodes was merged the "SA master node" seems to always be object 32; when we create the first test file (`$first_object` in the ksh script) we get the very first free object which is 2:

```
root@linux:~# touch /testpool/file
root@linux:~# zdb -dd testpool/
Dataset testpool [ZPL], ID 54, cr_txg 1, 24.0K, 7 objects

    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
         0    6   128K    16K  13.0K     512    32K   10.94  DMU dnode
        -1    1   128K    512      0     512    512  100.00  ZFS user/group used
        -2    1   128K    512      0     512    512  100.00  ZFS user/group used
         1    1   128K    512     1K     512    512  100.00  ZFS master node
         2    1   128K    512      0     512    512    0.00  ZFS plain file
        32    1   128K    512      0     512    512  100.00  SA master node
        33    1   128K    512      0     512    512  100.00  ZFS delete queue
        34    1   128K    512      0     512    512  100.00  ZFS directory
        35    1   128K  1.50K     1K     512  1.50K  100.00  SA attr registration
        36    1   128K    16K     8K     512    32K  100.00  SA attr layouts
```

But with the multi-threaded object allocator it doesn't seem to be possible (or maybe it just more difficult) to re-allocate objects <64, the test case is basically a while(true) loop.

Given that the objective is to _"verify that 'zfs send' drills appropriate holes"_ we just need to verify we reallocated an object, which is why i've changed the condition to `<= object`: we could play it safe and also verify we correctly generate FREEOBJECT records between OBJECT records (`zstreamdump` + `awk`?)

Reply:
Got it, that makes sense.  The change in object id of the "SA master node" was a known side effect the multi-threaded allocator which wasn't expected to cause any problems.  It just means we may create a few holes, but this the normal state for any active filesystem so that's fine.  Your fix for the test case looks entirely reasonable to me.  We could add the zstreamdump check too but I don't think it's necessary.